ipa-core/src/protocol/ipa_prf/malicious_security/quadratic_proofs.rs

```diff
@@ -0,0 +1,889 @@
+        share_new_0
+            .iter_mut()
+            .zip(share.0.chunks(r_f))
+            .for_each(|(u, chunk)| {
+                lagrange_evaluation_precomputed(chunk, &base_r, std::slice::from_mut(u));
+            });
+        share_new_1
+            .iter_mut()
+            .zip(share.1.chunks(r_f))
+            .for_each(|(u, chunk)| {
+                lagrange_evaluation_precomputed(chunk, &base_r, std::slice::from_mut(u));
+            });
+        // reassign new `share` for next iteration
+        *share.0 = share_new_0;
+        *share.1 = share_new_1;
+        share.0.truncate(s);
+        share.1.truncate(s);
+    }
+
+    debug_assert!(share.0.len() < r_f);
+
+    // final shares, compute output
+    let mut output_0 = F::ONE;
+    let mut output_1 = F::ONE;
+
+    let (final_share_0, final_share_1) = {
+        let mut vec_0 = Vec::<F>::with_capacity(r_f);
+        let mut vec_1 = Vec::<F>::with_capacity(r_f);
+        // add prss mask
+        vec_0.push(*mask);
+        vec_1.push(*mask);
+        // copy rest
+        vec_0.extend(&*share.0);
+        vec_1.extend(&*share.1);
+        // fill u and v with zeros such that it is a multiple of the recursion factor
+        vec_0.extend(std::iter::repeat(F::ZERO).take(vec_0.len() % r_f));
+        vec_1.extend(std::iter::repeat(F::ZERO).take(vec_1.len() % r_f));
+        (vec_0, vec_1)
+    };
+
+    // precompute Lagrange base for `r`
+    let mut base_r = vec![vec![F::ONE; r_f]; 1usize];
+    compute_lagrange_base(
+        std::slice::from_ref(r_p.last().unwrap()),
+        &e_p[0..r_f],
+        &denominators,
+        &mut base_r,
+    );
+
+    lagrange_evaluation_precomputed(&final_share_0, &base_r, std::slice::from_mut(&mut output_0));
+    lagrange_evaluation_precomputed(&final_share_1, &base_r, std::slice::from_mut(&mut output_1));
+
+    return (output_0, output_1);
+}
+
+/// computes `summand_of_g = p*q` and adds it to `g`
+/// polynomials `p` and `q` are represented by value `u`, `v`
+/// which are the evaluations of `p,q` at `evaluation_points[0..recursion_factor]`
+/// `g` is represented by its evaluation at `evaluation_points`
+fn compute_summand_of_g<F>(u: &[F], v: &[F], r_f: usize, base: &Vec<Vec<F>>, g: &mut [F]) -> ()
+where
+    F: Field + Invert,
+{
+    // check lengths
+    debug_assert_eq!(u.len(), v.len());
+    debug_assert_eq!(base.len(), u.len());
+
+    // define summand of g, g = sum of summand_of_g
+    // summand_of_g is represented by 2*recursion_factor many points.
+    let mut summand_of_g = vec![F::ONE; 2 * r_f];
+
+    // compute first recursion factor many points of summand_of_g
+    summand_of_g
+        .iter_mut()
+        .take(r_f)
+        .enumerate()
+        .for_each(|(i, x)| *x = u[i] * v[i]);
+
+    // compute summand_of_g = 1*u for rest of points
+    lagrange_evaluation_precomputed(u, &base, &mut summand_of_g[r_f..2 * r_f]);
```

Comment:
You should wrap the precomputed lagrange coefficients into a struct.

Given the way that the precomputed items are applied, you could have this in the form:

```rust
g[r_f..2*r_f] = base.evaluate(u) * g[r_f..2*r_f];
```

(This doesn't have to be Mul, as above, use a method on the returned value, if you prefer.)